### ISSUE TITLE
Add search results records method (to provide a scope instead of an array)

### ISSUE THREAD

**Issue by fabn** [ankane/searchkick#480]
It would be very useful to expose a `#records` method in returned results. You [already building a scope](https://github.com/ankane/searchkick/blob/master/lib/searchkick/results.rb#L147) for returned records. However you call `#to_a` on it and this trigger record loading.

Have a scope instead of an actual array has lot of benefits, some of them:
- Lazy loading, query won't be executed until you need results. This can be very useful to deal with cache keys, e.g. `<%= cache results.records.maximum(:updated_at) do %>`
- Results refinements (as you pointed in [this comment](https://github.com/ankane/searchkick/issues/140#issuecomment-36440598))
- Selective loading of retrieved fields. Suppose you have some big texts in database for returned results, one can use the awesome `pluck` method to not load all the other stuff. See the following example.

``` ruby
def autocomplete
  # This will only retrieve id and name attributes instead of whole records.
  results = Whatever.search(params[:query], fields: [{name: :text_start}], limit: 10)
  render json: results.records.pluck(:id, :name) # or even better results.pluck(:id, :name) if possible
end
```

**Comment by Mavvie** [comment]
The issue with implementing this is single table inheritance, specifically here https://github.com/ankane/searchkick/blob/master/lib/searchkick/results.rb#L37-L39

Say you have a `class Animal` with subclasses `Dog`, `Cat`, `Lizard`.

There's two solution on my mind:

1. Keep it as a relation of the parent class (`Animal`). This kinda sucks because you have to separate them yourself
2. Return a hash of the different types (`record['animal']`, `records['dog']`, etc). This kinda sucks if you want all of them together.

I'm leaning toward `1.`, any thoughts?

Here's an example of `2`: #489.

**Comment by fabn** [comment]
> Keep it as a relation of the parent class (Animal). This kinda sucks because you have to separate them yourself

I seldom used STI in rails, majority of times concerns are a better choice. 

In any case given you have a scope for `Animal` model imho the best choice is to return as is (i.e. `Animal.where(id: [list, of, returned, ids])`) because then the user can narrow the scope to suit its needs and once the scope is resolved (by doing the actual query) objects will be returned as subtypes, e.g.

```ruby
Animal.search('...').records.to_a.map(&:class) # => [Dog, Cat, Lizard]
```

Also if one uses STI usually there is some shared behaviors that is reusable across subclasses thus if one just needs animal names one can do `Animal.search(...).pluck(&:name).

Another option can be warn (or even raise) when the scope method is used on a table using STI.

**Comment by Mavvie** [comment]
Check out #490, it's pretty simple really. Just a WHERE id IN [results]

It does the query on the base class (in this case, `Animal`). It's pretty clean because I only implement it for ActiveRecord. I've never used Mongo or NoBrainer in my life so that's on someone else.

**Comment by Mavvie** [comment]
Like, I agree that STI isn't really something that can/should be handled in a `records` method like this. I think making it a relation of the base class makes a lot of sense.

**Comment by fabn** [comment]
Yes, I think that #490 it's all that we need. @ankane thoughts?

**Comment by ankane** [comment]
Merged #490.  However, the order of results will not be consistent with the Elasticsearch score, which is counterintuitive. For the case described above @fabn, I would use `load: false` to prevent an extra database round-trip.
